### PR TITLE
Sanitise: New transformation sub-package and some refactoring

### PR DESCRIPTION
**Pull request by mlange05**
This PR creates a new transformation sub-package for sanitisation and then refactors and improves several of the existing utilities for easier inclusion in external pipelines.

 In particular it separate associate-handling from sequence-association resolution, and adds a new utility for user-driven expression substitution. Each of those are now available as a single `Transformation` object and have been combined into a combined `SanitisePipeline`, each of which can now be used via external config constructors. The existing `SanitiseTransformation` is retained for backward compatibility.

In some more detail:
* Created a new sub-package `loki.transformations.sanitise` and moved according tests. In this sub-package, existing utilities for associates and sequence association have been separated
* Refactored sequence association resolver as an in-place `Transformer` to avoid awkward map handling. This requires some sub-classing in specific further use-case in `transformations.inline`.
* Provide `AssociateTransformation` to offer full or partial associate resolution together with the new associate-merging capabilities.
* Provide sequence association resolution as a standalone `Transformation`, which entails some method renaming
* Provide new `SubstituteExpressionTransformation` to expose string-based expression replacement
* Combine all of the above into a `SanitisePipeline`, which is intended as the long-term replacement for `SanitiseTransformation` - the latter, however, is retained for external backward compatibility
* Small tweaks to test (skip OMNI instead of xfail; it's faster and less messy in logs)

### CODE REVIEW

File: loki/transformations/sanitise/sequence_associations.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                n_dims = len(dummy.shape)
+                new_dims = []
+
+                if not arg.shape:
+                    # Hack: If we don't have a shape, short-circuit here
+                    new_dims = tuple(RangeIndex((None, None)) for _ in dummy.shape)
+                else:
+                    for s, lower in zip(arg.shape[:n_dims], arg.dimensions[:n_dims]):
+                        if isinstance(s, RangeIndex):
+                            new_dims += [RangeIndex((lower, s.stop))]
```
Comment (awnawab) [inline loki/transformations/sanitise/sequence_associations.py:123]: This is probably quite a common use case, and so could the tests please be extended to cover this?

File: loki/transformations/sanitise/sequence_associations.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        for dummy, arg in call.arg_map.items():
+            if check_if_scalar_syntax(arg, dummy):
+                found_scalar = True
+
+                n_dims = len(dummy.shape)
+                new_dims = []
+
+                if not arg.shape:
+                    # Hack: If we don't have a shape, short-circuit here
+                    new_dims = tuple(RangeIndex((None, None)) for _ in dummy.shape)
```
Comment (awnawab) [inline loki/transformations/sanitise/sequence_associations.py:119]: [no action] this line is also not covered by the testing, but I suspect we would only arrive here from incomplete enrichment of imported symbols, so probably no pressing need to extend the tests to cover this.

**Review by awnawab — CHANGES_REQUESTED** [review]
Many thanks for this cleanup! Having access to some of these utilities directly via scheduler configs will be very helpful indeed👌 I just have two small comments on extending the test coverage of the `SequenceAssociationTransformer`, one of which is optional but I feel the other is an important use case to cover.

File: loki/transformations/sanitise/sequence_associations.py (hunk 1 above)
Comment (mlange05) [reply loki/transformations/sanitise/sequence_associations.py:123]: Yes, it is and it was not covered in Santeri's original testing. I've slightly refactored his test and added specific cases for this. Please check if this now covers it.

File: loki/transformations/sanitise/sequence_associations.py (hunk 2 above)
Comment (mlange05) [reply loki/transformations/sanitise/sequence_associations.py:119]: Apologies, this one snuck in by accident (I needed it in one of the dev branches). I've now also added a test for this, so please see if this is ok.

**Review by awnawab — APPROVED** [review]
Thanks a lot for covering the missing lines, this is now good to go! 🙏

**Pull request merged**